cirq/circuits/circuit.py

```diff
@@ -792,6 +792,54 @@ def get_op_circuit(op: ops.Operation) -> Optional['cirq.FrozenCircuit']:
                     return False
         return True
 
+    def are_any_measurements_terminal(self):
+        """Whether any measurement gates are at the end of the circuit."""
+        return self.are_any_matches_terminal(protocols.is_measurement)
+
+    def are_any_matches_terminal(self, predicate: Callable[['cirq.Operation'], bool]):
+        """Check whether any of the ops that satisfy a predicate are terminal.
+
+        This method will transparently descend into any CircuitOperations this
+        circuit contains; as a result, it will misbehave if the predicate
+        refers to CircuitOperations. See the tests for an example of this.
+
+        Args:
+            predicate: A predicate on ops.Operations which is being checked.
+
+        Returns:
+            Whether or not any `Operation` s in a circuit that satisfy the
+            given predicate are terminal. Also checks within any CircuitGates
+            the circuit may contain.
+        """
+        from cirq.circuits import CircuitOperation
+
+        # TaggedOperations can wrap CircuitOperations.
+        def get_op_circuit(op: ops.Operation) -> Optional['cirq.FrozenCircuit']:
```

Comment:
I think we can do better in terms of deduping this and the `are_all_matches_terminal`, e.g. `get_op_circuit` could be pulled out to a shared private method. 
Not sure about the iteration over the moments is worth to dedupe but maybe

Reply:
Extracted `get_op_circuit`. Moment iteration probably isn't worth it, especially since the `any` version runs faster by iterating over the moments in reverse. I've reverse the `any` iterator in this commit.